File: Dsa-day3.cpp
class TrieNode {
public:
    unordered_map<string, TrieNode*> children;
    string name; 
    bool isDuplicate = false;  

    TrieNode(string name = "") : name(name) {}
};

class Solution {
public:
    unordered_map<string, int> subtreeCount;
    unordered_map<string, TrieNode*> subtreeMap;

    void insert(TrieNode* root, vector<string>& path) {
        TrieNode* node = root;
        for (const string& folder : path) {
            if (!node->children.count(folder)) {
                node->children[folder] = new TrieNode(folder);
            }
            node = node->children[folder];
        }
    }

    string serialize(TrieNode* node) {
        if (node->children.empty()) return "";

        vector<string> subs;
        for (auto& [name, child] : node->children) {
            string subSerial = serialize(child);
            subs.push_back("(" + name + subSerial + ")");
        }

        sort(subs.begin(), subs.end());  
        string serial = accumulate(subs.begin(), subs.end(), string(""));

        subtreeCount[serial]++;
        if (subtreeCount[serial] == 2) {
            subtreeMap[serial]->isDuplicate = true;
            node->isDuplicate = true;
        } else if (subtreeCount[serial] > 2) {
            node->isDuplicate = true;
        } else {
            subtreeMap[serial] = node;
        }

        return serial;
    }

    void collect(TrieNode* node, vector<string>& path, vector<vector<string>>& result) {
        for (auto& [name, child] : node->children) {
            if (!child->isDuplicate) {
                path.push_back(name);
                result.push_back(path);
                collect(child, path, result);
                path.pop_back();
            }
        }
    }

    vector<vector<string>> deleteDuplicateFolder(vector<vector<string>>& paths) {
        TrieNode* root = new TrieNode();
        for (auto& path : paths)
            insert(root, path);

        serialize(root);

        vector<vector<string>> result;
        vector<string> path;
        collect(root, path, result);

        return result;
    }
};

File: Dsa-day3.py
from collections import defaultdict
from typing import List

class TrieNode:
    def __init__(self, name=""):
        self.name = name
        self.children = {}
        self.is_duplicate = False

class Solution:
    def deleteDuplicateFolder(self, paths: List[List[str]]) -> List[List[str]]:
        root = TrieNode()
        
        def insert(path):
            node = root
            for folder in path:
                if folder not in node.children:
                    node.children[folder] = TrieNode(folder)
                node = node.children[folder]
        
        for path in paths:
            insert(path)

        subtree_count = defaultdict(int)
        serialization_map = {}

        def serialize(node):
            if not node.children:
                return ""

            serials = []
            for child_name in sorted(node.children):
                child_node = node.children[child_name]
                child_serial = serialize(child_node)
                serials.append(f"({child_name}{child_serial})")

            serial = "".join(serials)
            subtree_count[serial] += 1

            if subtree_count[serial] == 2:
                serialization_map[serial].is_duplicate = True
                node.is_duplicate = True
            elif subtree_count[serial] > 2:
                node.is_duplicate = True
            else:
                serialization_map[serial] = node

            return serial

        serialize(root)

        result = []

        def collect(node, path):
            for child_name, child in node.children.items():
                if not child.is_duplicate:
                    path.append(child_name)
                    result.append(path[:])
                    collect(child, path)
                    path.pop()

        collect(root, [])

        return result
